pennylane_lightning/src/algorithms/AdjointDiff.hpp

```diff
@@ -0,0 +1,617 @@
+void applyGeneratorRZ(SVType &sv, const std::vector<size_t> &wires,
+                      const bool adj = false) {
+    sv.applyOperation("PauliZ", wires, adj);
+}
+
+template <class T = double, class SVType = Pennylane::StateVector<T>>
+void applyGeneratorPhaseShift(SVType &sv, const std::vector<size_t> &wires,
+                              const bool adj = false) {
+    sv.applyOperation(getP11<T>(), wires, adj);
+}
+
+template <class T = double, class SVType = Pennylane::StateVector<T>>
+void applyGeneratorCRX(SVType &sv, const std::vector<size_t> &wires,
+                       const bool adj = false) {
+    const vector<size_t> internalIndices = sv.generateBitPatterns(wires);
+    const vector<size_t> externalWires = sv.getIndicesAfterExclusion(wires);
+    const vector<size_t> externalIndices =
+        sv.generateBitPatterns(externalWires);
+    for (const size_t &externalIndex : externalIndices) {
+        std::complex<T> *shiftedState = sv.getData() + externalIndex;
+        shiftedState[internalIndices[0]] = shiftedState[internalIndices[1]] = 0;
+        std::swap(shiftedState[internalIndices[2]],
+                  shiftedState[internalIndices[3]]);
+    }
+}
+
+template <class T = double, class SVType = Pennylane::StateVector<T>>
+void applyGeneratorCRY(SVType &sv, const std::vector<size_t> &wires,
+                       const bool adj = false) {
+    const vector<size_t> internalIndices = sv.generateBitPatterns(wires);
+    const vector<size_t> externalWires = sv.getIndicesAfterExclusion(wires);
+    const vector<size_t> externalIndices =
+        sv.generateBitPatterns(externalWires);
+    for (const size_t &externalIndex : externalIndices) {
+        std::complex<T> *shiftedState = sv.getData() + externalIndex;
+        std::complex<T> v0 = shiftedState[internalIndices[2]];
+        shiftedState[internalIndices[0]] = ZERO<T>();
+        shiftedState[internalIndices[1]] = ZERO<T>();
+        shiftedState[internalIndices[2]] =
+            -IMAG<T>() * shiftedState[internalIndices[3]];
+        shiftedState[internalIndices[3]] = IMAG<T>() * v0;
+    }
+}
+
+template <class T = double, class SVType = Pennylane::StateVector<T>>
+void applyGeneratorCRZ(SVType &sv, const std::vector<size_t> &wires,
+                       const bool adj = false) {
+    const vector<size_t> internalIndices = sv.generateBitPatterns(wires);
+    const vector<size_t> externalWires = sv.getIndicesAfterExclusion(wires);
+    const vector<size_t> externalIndices =
+        sv.generateBitPatterns(externalWires);
+    for (const size_t &externalIndex : externalIndices) {
+        std::complex<T> *shiftedState = sv.getData() + externalIndex;
+        shiftedState[internalIndices[0]] = shiftedState[internalIndices[1]] = 0;
+        shiftedState[internalIndices[3]] *= -1;
+    }
+}
+
+template <class T = double, class SVType = Pennylane::StateVector<T>>
+void applyGeneratorControlledPhaseShift(SVType &sv,
```

Comment:
Are these functions too simple to need doxygen?

Reply:
These were internal only to the translation unit, and will not be exposed to the user, so no need for Doxygen comments